tests/index.test.ts

```diff
@@ -0,0 +1,226 @@
+import { Pty } from '../index';
+import fs, { readdirSync, readlinkSync } from 'fs';
+import { describe, test, expect, onTestFinished, vi, beforeEach, afterEach } from 'vitest';
+
+const EOT = '\x04';
+const procSelfFd = '/proc/self/fd/';
+
+function createReadStreamFromPty(pty: Pty) {
+  const fd = pty.fd();
+  return fs.createReadStream('', { fd, autoClose: false });
+}
+
+function createWriteStreamToPty(pty: Pty) {
+  const fd = pty.fd();
+  return fs.createWriteStream('', { fd, autoClose: false });
+}
+
+const rejectOnNonEIO = (reject: (reason?: Error) => void) => (err: NodeJS.ErrnoException) => {
+  if (err.code && err.code.indexOf('EIO') !== -1) {
+    return;
+  }
+
+  reject(err);
+}
+
+type FdRecord = Record<string, string>;
+function getOpenFds(): FdRecord {
+  const fds: FdRecord = {};
+  if (process.platform !== 'linux') {
+    return fds;
+  }
+
+  for (const filename of readdirSync(procSelfFd)) {
+    try {
+      const linkTarget = readlinkSync(procSelfFd + filename);
+      if (linkTarget === 'anon_inode:[timerfd]') {
+        continue;
+      }
+
+      fds[filename] = linkTarget;
+    } catch (err: any) {
+      if (err.code === 'ENOENT') {
+        continue;
+      }
+      throw err;
+    }
+  }
+
+  return fds;
+}
+
+describe('PTY', () => {
+  let pty: Pty;
+  let oldFds: FdRecord;
+
+  beforeEach(() => {
+    oldFds = getOpenFds();
+  })
+
+  afterEach(() => {
+    pty.close();
+    expect(getOpenFds()).toStrictEqual(oldFds);
+  })
+
+  test('spawns and exits', () => new Promise<void>((done, reject) => {
+    const message = 'hello from a pty';
+    let buffer = '';
+
+    pty = new Pty({
+      command: '/bin/echo',
+      args: [message],
+      onExit: async (err, exitCode) => {
+        expect(err).toBeNull();
+        expect(exitCode).toBe(0);
+        vi.waitFor(() => expect(buffer.trim()).toBe(message));
+        done();
+      },
+    });
+
+    const readStream = createReadStreamFromPty(pty);
+    readStream.on('data', (chunk) => {
```

Comment:
hmm this is introducing a race :'( the problem here is that it is not _required_ for Node to consume the data immediately after the child sends it. in other words, the Node process can be chilling for quite a bit (i.e. the kernel it not forced to schedule it to run), the child process can send a bunch of data, and then later on this callback can be either invoked with an error (because `pty.close()` was already called) _or_ we will accidentally drop the data from the child

Reply:
this is the whole reason why we had the OG poll for both the child and the fd. i _think_ we need to keep that.

if we keep that, we can trust that the `onExit` will be called after the child is gonezo. now the last remaining thing will be to defer the `onExit` callback until after the read/write streams are done and we're :100: at that point!

Reply:
if the child dies/exits early the underlying fd should still be ok right because we only passed a `dup`'d fd to stdin/stdout/stderr

so shouldn't node be able to read the actual fd at its own leisure?